v19.1/advanced-client-side-transaction-retries.md

```diff
@@ -0,0 +1,70 @@
+
+1. When you use savepoints, you "hold your place in line" between attempts. Without savepoints, you're starting from scratch every time.
+2. Transactions increase their priority each time they're retried, increasing the likelihood they will succeed. This has a lesser effect than #1.
+
+## How transaction retries work
+
+A retryable transaction goes through the process described below, which maps to the following SQL statements:
+
+{% include copy-clipboard.html %}
+~~~ sql
+> BEGIN;                                  -- #1
+> SAVEPOINT cockroach_restart;            -- #2
+-- ... various transaction statements ... -- #3
+> RELEASE SAVEPOINT cockroach_restart;    -- #5 (Or #4, ROLLBACK, in case of retry error)
+> COMMIT;
+~~~
+
+1. The transaction starts with the [`BEGIN`](begin-transaction.html) statement.
+
+2. The [`SAVEPOINT`](savepoint.html) statement declares the intention to retry the transaction in the case of contention errors. Note that CockroachDB's savepoint implementation does not support all savepoint functionality, such as nested transactions. It must be executed after [`BEGIN`](begin-transaction.html) but before the first statement that manipulates a database.
+
+3. The statements in the transaction are executed.
+
+4. If a statement returns a retry error (identified via the `40001` error code or `"retry transaction"` string at the start of the error message), you can issue the [`ROLLBACK TO SAVEPOINT`](rollback-transaction.html) statement to restart the transaction and increase the transaction's priority. Alternately, the original [`SAVEPOINT`](savepoint.html) statement can be reissued to restart the transaction.
+
+    You must now issue the statements in the transaction again.
+
+    In cases where you do not want the application to retry the transaction, you can simply issue [`ROLLBACK`](rollback-transaction.html) at this point. Any other statements will be rejected by the server, as is generally the case after an error has been encountered and the transaction has not been closed.
+
+5. Once the transaction executes all statements without encountering contention errors, execute [`RELEASE SAVEPOINT`](release-savepoint.html) to commit the changes. If this succeeds, all changes made by the transaction become visible to subsequent transactions and are guaranteed to be durable if a crash occurs.
+
+    In some cases, the [`RELEASE SAVEPOINT`](release-savepoint.html) statement itself can fail with a retry error, mainly because transactions in CockroachDB only realize that they need to be restarted when they attempt to commit. If this happens, the retry error is handled as described in step 4.
+
+## Customizing the savepoint name
+
+{% include {{ page.version.version }}/misc/customizing-the-savepoint-name.md %}
+
+## Examples
```

Comment:
We talked earlier today in our PM 1:1 about prioritizing Go, Java, Python, and JavaScript as the languages of choice for application developer docs... Do we have a list of application-level features/areas that are covered by current example applications? Maybe by language? (e.g. "Advanced Client-Side Transaction Retries: Go [NA], Python [check], Java [check], Node [ ]"). I can see how that kind of list would be helpful when we start looking for ways to extend and improve the example application code that we currently have.

Reply:
I have a list in #5035, take a look and let me know what you think.  I think that list is just for this feature though.  If you mean "do we have a matrix-y thing that describes what features are covered by each example?" then I think the answer is no - at least I didn't find anything in [this quick search](https://github.com/cockroachdb/docs/issues?utf8=✓&q=is%3Aissue+is%3Aopen+code+sample).  Feel free to open a new issue!